fireworks/core/launchpad.py

```diff
@@ -1094,281 +633,443 @@ def detect_lostruns(self, expiration_secs=RUN_EXPIRATION_SECS, fizzle=False, rer
+                collection.
+            mongo (bool): spec_document uses mongo syntax to directly update the spec
+
+        TODO addres mongo variable, shouldn't be needed in general
         """
-        m_launch = self.get_launch_by_id(launch_id)
-        m_launch.launch_dir = launch_dir
-        self.launches.find_one_and_replace({'launch_id': m_launch.launch_id}, m_launch.to_db_dict(), upsert=True)
+        pass
 
-    def restore_backup_data(self, launch_id, fw_id):
+    @abstractmethod
+    def set_priority(self, fw_id: int, priority: int):
         """
-        For the given launch id and firework id, restore the back up data.
+        Set priority to the firework with the given id.
+
+        Args:
+            fw_id (int): firework id
+            priority (int): higher number -> higher priority
         """
-        if launch_id in self.backup_launch_data:
-            self.launches.find_one_and_replace({'launch_id': launch_id}, self.backup_launch_data[launch_id])
-        if fw_id in self.backup_fw_data:
-            self.fireworks.find_one_and_replace({'fw_id': fw_id}, self.backup_fw_data[fw_id])
+        pass
 
-    def complete_launch(self, launch_id, action=None, state='COMPLETED'):
+    @abstractmethod
+    def get_fw_ids(self, query: Dict=None, sort: Optional[List[Tuple[str,str]]] =None,
+                   limit: int=0, count_only: bool=False) -> List[int]:
         """
-        Internal method used to mark a Firework's Launch as completed.
+        Return all the fw ids that match a query.
 
         Args:
-            launch_id (int)
-            action (FWAction): the FWAction of what to do next
-            state (str): COMPLETED or FIZZLED
+            query (dict): representing a Mongo query
+            sort [(str,str)]: sort argument in Pymongo format
+            limit (int): limit the results
+            count_only (bool): only return the count rather than explicit ids
+
+        Returns:
+            list: list of firework ids matching the query
+        """
+        pass
+
+    @abstractmethod
+    def get_wf_ids(self, query: Dict=None, sort: Optional[List[Tuple[str,str]]] =None,
+                   limit: int=0, count_only: bool=False) -> List[int]:
+        """
+        Return one fw id for all workflows that match a query.
+
+        Args:
+            query (dict): representing a Mongo query
+            sort [(str,str)]: sort argument in Pymongo format
+            limit (int): limit the results
+            count_only (bool): only return the count rather than explicit ids
+
+        Returns:
+            list: list of firework ids
+        """
+        pass
+
+    @abstractmethod
+    def tuneup(self, bkground: bool=True):
+        """
+        Database tuneup: build indexes
+        """
+        pass
+
+    # might be able to define functions below here
+
+    @abstractmethod
+    def add_offline_run(self, fw_id: int, name: str):
```

Comment:
Added feature in 79ddf18209159adc50d750ef20465e07a14cdaee. Instead of just OFFLINE, the states are subdivided into OFFLINE-RESERVED, OFFLINE-READY, and (possible non-useful) OFFLINE-FIZZLED so that the state can be updated periodically while still being "offline". The offline states have the same rank as the online versions of the same state.